File: src/programmers/PRO_L3_등굣길.cpp
// dp 등굣길
#include <iostream>
#include <vector>

using namespace std;

int solution(int m, int n, vector<vector<int>> puddles) {
    
    int map[101][101] = {0};
    int dp[101][101] = {0};

    for (auto v: puddles) {
        map[v[1]][v[0]] = -1;
    }
    
    dp[1][0] = 1;
    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= m; j++) {
            if (map[i][j] == -1) {
                continue;
            } else {
                dp[i][j] = (dp[i-1][j] + dp[i][j-1]) % 1000000007;
            }
        }
    }
    
    int answer = dp[n][m];
    return answer;
}

int main() {
    int m = 4, n = 3;
    vector<vector<int>> puddles;
    puddles.push_back(vector<int>());
    puddles[0].push_back(2);
    puddles[0].push_back(2);

    cout << solution(4, 3, puddles);
    return 0;
}

File: src/programmers/PRO_L3_등굣길.java
package programmers;

/**
 * @{link} 	https://programmers.co.kr/learn/courses/30/lessons/42898
 * @date   	2020-05-09
 * @author 	rkddlsgur983
 * @idea	동적 계획법
 */
public class PRO_L3_등굣길 {
	private int mod = 1000000007;
	public int solution(int m, int n, int[][] puddles) {
        int[][] map = new int[n][m];
        for (int[] p: puddles) {
        	map[p[1]-1][p[0]-1] = -1;
        }
        for (int i = 0; i < m; ++i) {
        	if (map[0][i] == -1) break;
        	map[0][i] = 1;
        }
        for (int i = 0; i < n; ++i) {
        	if (map[i][0] == -1) break;
        	map[i][0] = 1;
        }
        for (int i = 1; i < n; ++i) {
        	for (int j = 1; j < m; ++j) {
        		if (map[i][j] == -1) continue;
        		if (map[i-1][j] != -1) {
        			map[i][j] = (map[i][j] + map[i-1][j])%mod;
        		}
        		if (map[i][j-1] != -1) {
        			map[i][j] = (map[i][j] + map[i][j-1])%mod;
        		}
        	}
        }
		int answer = map[n-1][m-1];
        return answer;
    }
}
